### PR TITLE
Implement \ref backreferences

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: bootstrap/parser.go [hunk 1]
```diff
@@ -242,15 +243,70 @@ type seqParser struct {
...
 	furthest := *input
-	for _, parser := range p.parsers {
+
+	for _, item := range p.parsers {
 		var v interface{}
-		if err := parser.Parse(input, &v); err != nil {
-			*input = furthest
-			return err
+		if ref, ok := item.(*REF); ok {
+			for i := 0; i < len(result); i++ {
```
Comment (marcelocantos) [inline bootstrap/parser.go]:
```
for i, v := range result {
```

File: examples/xml.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+xml  -> "<" NAME attr* ">" xml?  "</" \NAME ">" | CDATA=/{[^<]*};
```
Comment (marcelocantos) [inline examples/xml.txt]: Change `*` to `+` in the CDATA regexp.

**Review by marcelocantos — CHANGES_REQUESTED** [review]

File: bootstrap/grammar.go [hunk 3]
```diff
@@ -51,9 +52,10 @@ named   -> (IDENT op="=")? atom;
 quant   -> op=/{[?*+]}
          | "{" min=INT? "," max=INT? "}"
          | op=/{<:|:>?} opt_leading=","? named opt_trailing=","?;
-atom    -> IDENT | STR | RE | "(" term ")" | "(" ")";
+atom    -> IDENT | STR | RE | REF | "(" term ")" | "(" ")";
 
 // Terminals
+REF		-> ‵\‵ IDENT;
```
Comment (marcelocantos) [inline bootstrap/grammar.go]: nit: Better in the same of order of appearance as in the non-terminals section. On the same basis, if you feel like it, move COMMENT to the front.

File: bootstrap/grammar.go [hunk 4]
```diff
@@ -100,9 +102,10 @@ var grammarGrammar = Grammar{
 		},
 	},
 	named: Seq{Opt(Seq{ident, Eq("op", S("="))}), atom},
-	atom:  Oneof{ident, str, re, Seq{S("("), term, S(")")}, Seq{S("("), S(")")}},
+	atom:  Oneof{ident, str, re, ref, Seq{S("("), term, S(")")}, Seq{S("("), S(")")}},
 
 	// Terminals
+	ref:     Seq{S("\\"), ident},
```
Comment (marcelocantos) [inline bootstrap/grammar.go]: nit: As above, keep ordering consistent.

File: bootstrap/parser.go (hunk 1 above)
Comment (ghost) [reply bootstrap/parser.go]: i think i intended to iterate through the results slice backwards, not sure if thats needed, switching.

File: examples/xml.txt (hunk 2 above)
Comment (ghost) [reply examples/xml.txt]: awesome

File: bootstrap/grammar.go (hunk 3 above)
Comment (ghost) [reply bootstrap/grammar.go]: k

File: bootstrap/grammar.go (hunk 4 above)
Comment (ghost) [reply bootstrap/grammar.go]: k

**Pull request merged**